docs/en/03_Upgrading/01_Upgrading_project.md

```diff
@@ -0,0 +1,1188 @@
+        "branch-alias": {
+            "4.x-dev": "4.2.x-dev"
+        },
+        "expose": [
+            "app/scripts",
+            "app/images",
+            "app/css"
+        ]
+    },
+    // ...
+}
+```
+
+For the change to take affect, run the following command: `composer vendor-expose`.
+
+### Referencing static assets in your PHP code
+Wherever you would have use an hardcoded path, you can now use the `path/to/file.css` syntax. To reference a static file from a module, prefix the path with the module's name (e.g.: `silverstripe/admin:client/dist/js/bundle.js`).
+
+To add some javascript and css files to your requirements from your PHP code, you could use this syntax:
+```php
+use SilverStripe\View\Requirements;
+
+# Load your own style and scripts
+Requirements::css('app/css/styles.css');
+Requirements::script('app/scripts/client.css');
+
+# Load some assets from a module.  
+Requirements::script('silverstripe/blog: js/main.bundle.js');
+```
+
+You can `SilverStripe\Core\Manifest\ModuleResourceLoader` to get the web path of file.
+```php
+ModuleResourceLoader::singleton()->resolveURL('app/images/road-runner.jpg');
+ModuleResourceLoader::singleton()->resolveURL('silverstripe/blog: js/main.bundle.js');
+```
+
+You can use `SilverStripe\View\ThemeResourceLoader` to access files from your theme:
+```php
+$cssResourcePath = ThemeResourceLoader::inst()->findThemedResource('css/layout.css');
+$relativeUrl = ModuleResourceLoader::singleton()->resolveURL($cssResourcePath);
+```
+
+For classes that expect icons, you can specify theme with:
+```php
+class ListingPage extends \Page
+{
+    private static $icon = 'app/images/sitetree_icon.png';
+}
+
+class MyCustomModelAdmin extends \SilverStripe\Admin\ModelAdmin
+{
+    private static $menu_icon = 'app/images/modeladmin_icon.png';
+}
+```
+
+### Referencing static assets in template files
+SS template files accept a similar format for referencing static assets. Go through your assets files and remove hardcoded references.
+
+```html
+<img src="$resourceURL(app/images/coyote.png)" />
+<% require css("app: css/styles.css") %>
+```
+
+### Finalising removal of hardcoded paths for referencing static assets
+
+All your assets should be loading properly now.
+
+This is a good point to commit your changes to your source control system before moving on to the next step.
+
+
+## Step 10 - Update database class references {#step10}
+
+If you've updated your class names to use namespaces you will need to reflect those changes in any existing database fields. For example, if you've renamed your `HomePage` class to `App\HomePage` then the database `ClassName` column needs to be updated to point to the `App\HomePage` class, otherwise the CMS will tell you that the page is obsolete. This also applies to polymorphic relationships.
+
+There is no automated way to do this, but you can use the list generated in .upgrade.yml and copy it to `app/_config/legacy.yml`, removing any classes that don't extend DataObject.
```

Comment:
Is automating this planned? Seems possible...

Reply:
I've added an issue for it https://github.com/silverstripe/silverstripe-upgrader/issues/158

Reply:
We've got other issues related to `classname_value_remapping`:
* https://github.com/silverstripe/silverstripe-framework/issues/8344
* https://github.com/silverstripe/silverstripe-framework/issues/8686